Mathlib/Algebra/Group/AddChar.lean

```diff
@@ -157,36 +169,77 @@ def toAddMonoidHomEquiv : AddChar A M ≃ (A →+ Additive M) where
+    toAddMonoidHomEquiv ψ a = Additive.ofMul (ψ a) := rfl
+
+@[simp] lemma toAddMonoidHomEquiv_symm_apply (ψ : A →+ Additive M) (a : A) :
+    toAddMonoidHomEquiv.symm ψ a = Additive.toMul (ψ a) := rfl
+
 /-- The trivial additive character (sending everything to `1`) is `(1 : AddChar A M).` -/
-instance instOne : One (AddChar A M) := (toMonoidHomEquiv A M).one
+instance instOne : One (AddChar A M) := toMonoidHomEquiv.one
 
--- Porting note: added
+@[simp, norm_cast] lemma coe_one : ⇑(1 : AddChar A M) = 1 := rfl
 @[simp] lemma one_apply (a : A) : (1 : AddChar A M) a = 1 := rfl
 
 instance instInhabited : Inhabited (AddChar A M) := ⟨1⟩
 
 /-- Composing a `MonoidHom` with an `AddChar` yields another `AddChar`. -/
 def _root_.MonoidHom.compAddChar {N : Type*} [Monoid N] (f : M →* N) (φ : AddChar A M) :
-    AddChar A N :=
-  (toMonoidHomEquiv A N).symm (f.comp φ.toMonoidHom)
+    AddChar A N := toMonoidHomEquiv.symm (f.comp φ.toMonoidHom)
 
-@[simp]
+@[simp, norm_cast]
 lemma _root_.MonoidHom.coe_compAddChar {N : Type*} [Monoid N] (f : M →* N) (φ : AddChar A M) :
     f.compAddChar φ = f ∘ φ :=
   rfl
 
+@[simp, norm_cast]
+lemma _root_.MonoidHom.compAddChar_apply (f : M →* N) (φ : AddChar A M) : f.compAddChar φ = f ∘ φ :=
+  rfl
+
+lemma _root_.MonoidHom.compAddChar_injective_left (ψ : AddChar A M) (hψ : Surjective ψ) :
+    Injective fun f : M →* N ↦ f.compAddChar ψ := by
+  rintro f g h; rw [DFunLike.ext'_iff] at h ⊢; exact hψ.injective_comp_right h
+
+lemma _root_.MonoidHom.compAddChar_injective_right (f : M →* N) (hf : Injective f) :
+    Injective fun ψ : AddChar B M ↦ f.compAddChar ψ := by
+  rintro ψ χ h; rw [DFunLike.ext'_iff] at h ⊢; exact hf.comp_left h
+
 /-- Composing an `AddChar` with an `AddMonoidHom` yields another `AddChar`. -/
-def compAddMonoidHom {B : Type*} [AddMonoid B] (φ : AddChar B M) (f : A →+ B) : AddChar A M :=
-  (toAddMonoidHomEquiv A M).symm (φ.toAddMonoidHom.comp f)
+def compAddMonoidHom (φ : AddChar B M) (f : A →+ B) : AddChar A M :=
+  (toAddMonoidHomEquiv).symm (φ.toAddMonoidHom.comp f)
+
+@[simp, norm_cast]
+lemma coe_compAddMonoidHom (φ : AddChar B M) (f : A →+ B) : φ.compAddMonoidHom f = φ ∘ f := rfl
+
+@[simp] lemma compAddMonoidHom_apply (ψ : AddChar B M) (f : A →+ B)
+    (a : A) : ψ.compAddMonoidHom f a = ψ (f a) := rfl
+
+lemma compAddMonoidHom_injective_left (f : A →+ B) (hf : Surjective f) :
+    Injective fun ψ : AddChar B M ↦ ψ.compAddMonoidHom f := by
+  rintro ψ χ h; rw [DFunLike.ext'_iff] at h ⊢; exact hf.injective_comp_right h
 
-@[simp] lemma coe_compAddMonoidHom {B : Type*} [AddMonoid B] (φ : AddChar B M) (f : A →+ B) :
-    φ.compAddMonoidHom f = φ ∘ f := rfl
+lemma compAddMonoidHom_injective_right (ψ : AddChar B M) (hψ : Injective ψ) :
+    Injective fun f : A →+ B ↦ ψ.compAddMonoidHom f := by
+  rintro f g h; rw [DFunLike.ext'_iff] at h ⊢; exact hψ.comp_left h
+
+lemma eq_one_iff : ψ = 1 ↔ ∀ x, ψ x = 1 := DFunLike.ext_iff
+lemma ne_one_iff : ψ ≠ 1 ↔ ∃ x, ψ x ≠ 1 := DFunLike.ne_iff
```

Comment:
```suggestion
lemma eq_one_iff : ψ = 1 ↔ ∀ x, ψ x = 1 := DFunLike.ext_iff

lemma ne_one_iff : ψ ≠ 1 ↔ ∃ x, ψ x ≠ 1 := DFunLike.ne_iff
```

Reply:
I'm not convinced about this one. Putting the two next to each other shows that they are completely analogous. This is done in many place throughout mathlib

Reply:
Fair enough